[FZ Editor] Add test to editor #197 + feedback

garv5014 opened the pull request:
<!-- Enter a brief description/summary of your PR here. What does it fix/what does it change/how was it tested (even manually, if necessary)? -->
## Summary of the Pull Request
This pull request is to add a tests to the FancyZoneEditor as is requested in issue [#197](https://github.com/microsoft/PowerToys/issues/197). We added a new msTest project and lightly tested two of the existing models (GridLayoutModel.cs and DefaultLayoutModel.cs). We aimed to test existing observed behavior. 
<!-- Please review the items on the PR checklist before submitting-->
## PR Checklist

- [x] **Closes:** We don't expect the issue to be closed off of this on pr. [#197](https://github.com/microsoft/PowerToys/issues/197)
- [ ] **Communication:** I've discussed this with core contributors already. If work hasn't been agreed, this work might be rejected
- [x] **Tests:** Added/updated and all pass
- [x] **Localization:** All end user facing strings can be localized
- [ ] **Dev docs:** Added/updated
- [x] **New binaries:** Added on the required places
   - [x] [JSON for signing](https://github.com/microsoft/PowerToys/blob/main/.pipelines/ESRPSigning_core.json) for new binaries
   - [x] [WXS for installer](https://github.com/microsoft/PowerToys/blob/main/installer/PowerToysSetup/Product.wxs) for new binaries and localization folder
   - [x] [YML for CI pipeline](https://github.com/microsoft/PowerToys/blob/main/.pipelines/ci/templates/build-powertoys-steps.yml) for new test projects
   - [x] [YML for signed pipeline](https://github.com/microsoft/PowerToys/blob/main/.pipelines/release.yml)
- [ ] **Documentation updated:** If checked, please file a pull request on [our docs repo](https://github.com/MicrosoftDocs/windows-uwp/tree/docs/hub/powertoys) and link it here: #xxx

<!-- Provide a more detailed description of the PR, other things fixed or any additional comments/features here -->
## Detailed Description of the Pull Request / Additional comments
We made tests for the GridLayoutModel that test the IsModelValid() method.  We also tested that the DefaultLayoutModel can only have one default layout for each layout type set at a time. While testing this section of the code we found a temporal coupling in the code that we would also like feedback on. [Temporal Coupling](https://github.com/microsoft/PowerToys/issues/29050)
We would like feedback on the testing that has been done so far to make sure we are going about it correctly. We don't intend for this pr to close 197
<!-- Describe how you validated the behavior. Add automated tests wherever possible, but list manual validation steps taken as well -->
## Validation Steps Performed
See tests

garv5014 commented:
> @garv5014 please read the following Contributor License Agreement(CLA). If you agree with the CLA, please reply with the following information.
> 
> ```
> @microsoft-github-policy-service agree [company="{your company}"]
> ```
> 
> > Options:
> > 
> > * (default - no company specified) I have sole ownership of intellectual property rights to my Submissions and I am not making Submissions in the course of work for my employer.
> > 
> > ```
> > @microsoft-github-policy-service agree
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > 
> > * (when company given) I am making Submissions in the course of work for my employer (or my employer has intellectual property rights in my Submissions by contract or applicable law). I have permission from my employer to make Submissions and enter into this Agreement on behalf of my employer. By signing below, the defined term “You” includes me and my employer.
> > 
> > ```
> > @microsoft-github-policy-service agree company="Microsoft"
> > ```
> 
> Contributor License Agreement

@microsoft-github-policy-service agree

crutkas commented on UnitTestFancyZonesEditor/UnitTestFancyZonesEditor.csproj:
```diff
@@ -0,0 +1,29 @@
+    <PackageReference Include="Microsoft.NET.Test.Sdk" />
+    <PackageReference Include="MSTest.TestAdapter" />
+    <PackageReference Include="MSTest.TestFramework" />
+    <PackageReference Include="coverlet.collector" />
```
https://github.com/coverlet-coverage/coverlet

crutkas replied:
this is also causing the build to fail due to the notice file

garv5014 replied:
So we should remove the coverlet-coverage package from the test?

crutkas replied:
If we don't need it, don't include it  :)

I have a powershell script verify all nuget packages we leverage in the notice.md.  If anything changes (version, a new one, one removed), it blocks the PR until the change is reflected.  Too easy of a thing for us to miss and 100% can be caught via CI

garv5014 replied:
Gotcha makes sense. Thanks for the quick responses!

SeraphimaZykova commented on UnitTestFancyZonesEditor/GridLayoutModelTests.cs:
```diff
@@ -0,0 +1,95 @@
+
+using FancyZonesEditor.Models;
+
+namespace UnitTestFancyZonesEditor;
```
Different namespaces in `GridLayoutModelTests` and `DefaultLayoutssModelTests`, let's keep one of them.

SeraphimaZykova commented on src/modules/fancyzones/UnitTests-FancyZonesEditor/UnitTests-FancyZonesEditor.csproj:
```diff
@@ -0,0 +1,28 @@
+    <IsPackable>false</IsPackable>
+    <IsTestProject>true</IsTestProject>
+  </PropertyGroup>
+
```
Let's add an output path as in other test projects:
```suggestion

  <PropertyGroup>
  <OutputPath>..\..\..\..\$(Platform)\$(Configuration)\tests\UnitTest-FancyZonesEditor\</OutputPath>
  </PropertyGroup>
  
```

SeraphimaZykova commented on UnitTestFancyZonesEditor/UnitTestFancyZonesEditor.csproj:
Let's remove this file, it duplicates `src\modules\fancyzones\UnitTests-FancyZonesEditor\UnitTests-FancyZonesEditor.csproj`.

SeraphimaZykova commented on UnitTestFancyZonesEditor/Usings.cs:
Let's remove this file, it duplicates `src\modules\fancyzones\UnitTests-FancyZonesEditor\Usings.cs`.

SeraphimaZykova commented on UnitTestFancyZonesEditor/GridLayoutModelTests.cs:
Let's remove this file, it duplicates `src\modules\fancyzones\UnitTests-FancyZonesEditor\GridLayoutModelTests.cs`.

SeraphimaZykova commented on src/codeAnalysis/GlobalSuppressions.cs:
It's better to keep the final commit clean and not include unrelated files. Please revert the changes made in this file.

SeraphimaZykova commented on src/modules/fancyzones/UnitTests-FancyZonesEditor/DefaultLayoutsModelTests.cs:
```diff
@@ -0,0 +1,71 @@
+[TestClass]
+public class DefaultLayoutsModelTests
+{
+    [TestMethod]
+    public void SettingFirstLayoutShouldBeLengthOne()
+    {
+        var defaultLayoutsModel = new DefaultLayoutsModel();
+        GridLayoutModel gridLayoutModel = new GridLayoutModel();
+
+        defaultLayoutsModel.Set(gridLayoutModel, MonitorConfigurationType.Horizontal);
+
+        var actualCount = defaultLayoutsModel.Layouts.Count;
+        Assert.AreEqual(1, actualCount);
+    }
```
I don't clearly understand the purpose of this test.

garv5014 replied:
This test was to make sure that when you set one of the defaults that layouts count did increase. Showing that something was added to the layouts. This was a little more of a stepping stone test to make sure we understood the way the models were being used. Can be removed if deemed appropriate

SeraphimaZykova reviewed (changes requested):
First of all, thank you for your work. It's a good start to covering FZEditor with tests.
I've left some comments for you regarding some changes that need to be made, mostly centered around keeping the project clean and consistent. Thank you!

SeraphimaZykova commented on src/modules/fancyzones/UnitTests-FancyZonesEditor/GridLayoutModelTests.cs:
```diff
@@ -0,0 +1,95 @@
+
+using FancyZonesEditor.Models;
+
+namespace UnitTestFancyZonesEditor;
```
```suggestion
namespace UnitTestsFancyZonesEditor;
```

The last missing change. They still have different namespaces.
 
![image](https://github.com/microsoft/PowerToys/assets/8949536/a76b3fdd-8858-438e-a809-8f149a6e6536)

SeraphimaZykova reviewed (approved):
LGTM!

jaimecbernardo reviewed (changes requested):
Hi,
Thanks for the contribution.

I've just given this PR a look and it seems it's not really running in CI yet.
For that you'd have to add
```
      **\UnitTests-FancyZonesEditor.dll
```
in the Native dlls portion of `.pipelines/ci/templates/build-powertoys-steps.yml`. I've tried to do it, but it seems like you've blocked writes to your fork from contributors when you've opened the PR.

I'd like to bring this in only after the tests run in CI as well, so can you please add that entry so we can verify if they pass CI as well? Thanks you.

garv5014 commented:
> Hi, Thanks for the contribution.
> 
> I've just given this PR a look and it seems it's not really running in CI yet. For that you'd have to add
> 
> ```
>       **\UnitTests-FancyZonesEditor.dll
> ```
> 
> in the `'MS Tests'` portion of `.pipelines/ci/templates/build-powertoys-steps.yml`. I've tried to do it, but it seems like you've blocked writes to your fork from contributors when you've opened the PR.
> 
> I'd like to bring this in only after the tests run in CI as well, so can you please add that entry so we can verify if they pass CI as well? Thank you.
> 
> Edit: Corrected that it should be in the `'MS Tests'` tests part.

Added to the pipeline. Is is standard practice to give original collaborators access to forked projects? I am assuming that is why you couldn't change it because the PR came from a forked project

garv5014 commented:
> Hi, Thanks for the contribution.
> 
> I've just given this PR a look and it seems it's not really running in CI yet. For that you'd have to add
> 
> ```
>       **\UnitTests-FancyZonesEditor.dll
> ```
> 
> in the `'MS Tests'` portion of `.pipelines/ci/templates/build-powertoys-steps.yml`. I've tried to do it, but it seems like you've blocked writes to your fork from contributors when you've opened the PR.
> 
> I'd like to bring this in only after the tests run in CI as well, so can you please add that entry so we can verify if they pass CI as well? Thank you.
> 
> Edit: Corrected that it should be in the `'MS Tests'` tests part.

We added the tests but then some other test what we didn't make failed on the run. I don't think us adding that dll line could have cause it? Thoughts?

crutkas commented:
@jaimecbernardo looks like one of the SVG tests is flaky the failures here weren't due to any change

jaimecbernardo commented:
> @jaimecbernardo looks like one of the SVG tests is flaky the failures here weren't due to any change

Yes, WebView 2 initialization is flaky in CI 🤷

jaimecbernardo commented:
> > Hi, Thanks for the contribution.
> > I've just given this PR a look and it seems it's not really running in CI yet. For that you'd have to add
> > ```
> >       **\UnitTests-FancyZonesEditor.dll
> > ```
> > 
> > 
> > in the `'MS Tests'` portion of `.pipelines/ci/templates/build-powertoys-steps.yml`. I've tried to do it, but it seems like you've blocked writes to your fork from contributors when you've opened the PR.
> > I'd like to bring this in only after the tests run in CI as well, so can you please add that entry so we can verify if they pass CI as well? Thank you.
> > Edit: Corrected that it should be in the `'MS Tests'` tests part.
> 
> Added to the pipeline. Is is standard practice to give original collaborators access to forked projects? I am assuming that is why you couldn't change it because the PR came from a forked project

It's possible to add commits by the projects contributor by using this checkbox when creating the PR 😄 
![image](https://github.com/microsoft/PowerToys/assets/26118718/428ead41-7964-492b-9f71-139f5601965d)

jaimecbernardo reviewed (approved):
LGTM! Thank you for the contribution!
